### ISSUE TITLE
Version >= 0.32.0. Avoid comma after the last item of a hash, unless each item is on its own line

### ISSUE THREAD

**Issue by vinagrito** [bbatsov/rubocop#2403]
My current
```
0.35.0 (using Parser 2.2.3.0, running on ruby 2.2.3 x86_64-darwin15)
```
I'm having this:
```ruby
1  def some_method
2   super.merge({
3      a_key: {
4       another_key: "value",
5     },
6   },)
7 end
```
I noticed that since `0.32.0` released on <b>June 6, 2015</b> the next syntax enforces:
Having the comma on <b>line 6</b> seems a bit off.

Now, with the latest version the comma on line 4 is no longer encouraged.

Or am I missing something?

**Comment by bbatsov** [comment]
We decided this doesn't make much sense for method arguments and disabled
it by default in 0.35.

**Comment by vinagrito** [comment]
I don't see <b>Line 4</b> as method argument. I mean even this:
```ruby
def request_params
 {
   id: event.id,
 }
end
# FAILS
```
Again about <b>line 6</b> -  The only way I found it be pass for me is:
```ruby
1  def some_method
2   super.merge(
3     a_key: {
4       another_key: "value"
5     },
6   )
7 end
# PASSES 
```
(removed the brackets that enclose the hash passed into `#merge`)

Sorry if I haven't it 100% but <b>Line 4</b> should still enforce that comma

**Comment by bbatsov** [comment]
I'll have to check the default config. Maybe we change it or something. AFAIK we've never required trailing commas by default.

**Comment by alexdowad** [comment]
@vinagrito, I don't understand what the problem is here. RC doesn't like an unneeded trailing comma in a hash. Makes sense to me. Do you have a good argument why it should be different?

**Comment by vinagrito** [comment]
@alexdowad in my second example i.e  `a_key: { another_key: "value" }` - that's a hash as well, yet the comma there seems to be fine

**Comment by alexdowad** [comment]
Could you try this again after my open PR is merged and see if the behavior is as desired?

**Comment by vinagrito** [comment]
Sure, can you please link me to it?

**Comment by alexdowad** [comment]
Just waiting for review. I will let you know.

**Comment by alexdowad** [comment]
@vinagrito Can you please try the newest `master` now?

**Comment by vinagrito** [comment]
@alexdowad, just run the `master` version and indeed it now complains about the comma everywhere :). 
So 3 things here. 
1. Good that it is consistent :+1: 
2. Why was it decided to make this change?. Before, when pushing to github you wouldn't see the diff with a single comma added to the end of the line
3. My initial answer had only two points :)... I noticed that I get some warnings (which is fine) when running `bundle exec rubocop spec/requests/some_dir/some_file_spec.rb`. The problem is that usually warnings are not a show stopper. 
```
Warning: unrecognized cop TrailingComma found in /Users/daniel/work/stars/.rubocop.yml
Warning: unrecognized cop SingleSpaceBeforeFirstArg found in /Users/daniel/work/stars/.rubocop.yml
Warning: unrecognized cop Rails/DefaultScope found in /Users/daniel/work/stars/.rubocop.yml
```
when leaving this lines in the rubocop.yml file then it would throw
```
➜  stars git:(master) ✗ bundle exec rubocop spec/requests/private/v1/accounts/update_spec.rb
Warning: unrecognized cop TrailingComma found in /Users/daniel/work/stars/.rubocop.yml
Warning: unrecognized cop SingleSpaceBeforeFirstArg found in /Users/daniel/work/stars/.rubocop.yml
Warning: unrecognized cop Rails/DefaultScope found in /Users/daniel/work/stars/.rubocop.yml
Inspecting 1 file


0 files inspected, no offenses detected
undefined method `to_sym' for {"name"=>"convention"}:Hash
/Users/daniel/work/stars/.bundle/bundler/gems/rubocop-cbbf7b949593/lib/rubocop/cop/severity.rb:28:in `name_from_code'
/Users/daniel/work/stars/.bundle/bundler/gems/rubocop-cbbf7b949593/lib/rubocop/cop/severity.rb:34:in `initialize'
/Users/daniel/work/stars/.bundle/bundler/gems/rubocop-cbbf7b949593/lib/rubocop/cop/offense.rb:61:in `new'
/Users/daniel/work/stars/.bundle/bundler/gems/rubocop-cbbf7b949593/lib/rubocop/cop/offense.rb:61:in `initialize'
/Users/daniel/work/stars/.bundle/bundler/gems/rubocop-cbbf7b949593/lib/rubocop/cached_data.rb:53:in `new'
/Users/daniel/work/stars/.bundle/bundler/gems/rubocop-cbbf7b949593/lib/rubocop/cached_data.rb:53:in `block in deserialize_offenses'
/Users/daniel/work/stars/.bundle/bundler/gems/rubocop-cbbf7b949593/lib/rubocop/cached_data.rb:49:in `map!'
/Users/daniel/work/stars/.bundle/bundler/gems/rubocop-cbbf7b949593/lib/rubocop/cached_data.rb:49:in `deserialize_offenses'
/Users/daniel/work/stars/.bundle/bundler/gems/rubocop-cbbf7b949593/lib/rubocop/cached_data.rb:14:in `from_json'
/Users/daniel/work/stars/.bundle/bundler/gems/rubocop-cbbf7b949593/lib/rubocop/result_cache.rb:70:in `load'
/Users/daniel/work/stars/.bundle/bundler/gems/rubocop-cbbf7b949593/lib/rubocop/runner.rb:84:in `process_file'
/Users/daniel/work/stars/.bundle/bundler/gems/rubocop-cbbf7b949593/lib/rubocop/runner.rb:59:in `block in inspect_files'
/Users/daniel/work/stars/.bundle/bundler/gems/rubocop-cbbf7b949593/lib/rubocop/runner.rb:57:in `each'
/Users/daniel/work/stars/.bundle/bundler/gems/rubocop-cbbf7b949593/lib/rubocop/runner.rb:57:in `inspect_files'
/Users/daniel/work/stars/.bundle/bundler/gems/rubocop-cbbf7b949593/lib/rubocop/runner.rb:35:in `run'
/Users/daniel/work/stars/.bundle/bundler/gems/rubocop-cbbf7b949593/lib/rubocop/cli.rb:30:in `run'
/Users/daniel/work/stars/.bundle/bundler/gems/rubocop-cbbf7b949593/bin/rubocop:14:in `block in <top (required)>'
/Users/daniel/.rbenv/versions/2.2.3/lib/ruby/2.2.0/benchmark.rb:303:in `realtime'
/Users/daniel/work/stars/.bundle/bundler/gems/rubocop-cbbf7b949593/bin/rubocop:13:in `<top (required)>'
/Users/daniel/work/stars/.bundle/bin/rubocop:23:in `load'
/Users/daniel/work/stars/.bundle/bin/rubocop:23:in `<main>'

**Comment by alexdowad** [comment]
> Why was it decided to make this change?. Before, when pushing to github you wouldn't see the diff with a single comma added to the end of the line

There have been a large number of changes made. Can you please specify exactly which you are talking about?

(Please note that RuboCop is a "work in progress". It has not yet reached its first stable release, and the developers are trying to improve it as quickly as possible. So there will be noticeable changes in every release.)

> I noticed that I get some warnings (which is fine) when running bundle exec rubocop spec/requests/some_dir/some_file_spec.rb

Thank you for drawing attention to this. We have removed some cops, and renamed others. I will add more specific error messages which will tell users exactly how to bring your configuration up-to-date.

> undefined method `to_sym' for {"name"=>"convention"}:Hash

It appears that stale cache data is being used. This shouldn't be the case...

**Comment by alexdowad** [comment]
Could you open `lib/rubocop/result_cache.rb`, and just after line 58, add something like:

```ruby
puts "***** RuboCop source checksum: #{rubocop_checksum.inspect}"
```

Then run RuboCop and see what it prints?

**Comment by vinagrito** [comment]
>There have been a large number of changes made. Can you please specify exactly which you are talking about?

I meant the comma change. I think there was a point where it was decided to have the comma after the last element in a hash. Which since version 0.34 (I think) has changed. I just want to know the motivation

>Could you open lib/rubocop/result_cache.rb, and just after line 58, add something like:
```ruby
puts "***** RuboCop source checksum: #{rubocop_checksum.inspect}"
```

***** RuboCop source checksum: "4729a0d9de363de50542ef0bf8f250b1"

I'm sorry to mix subjects...but since we're here. Where can I see a list of the existing cops after a new version comes out?

**Comment by alexdowad** [comment]
> I meant the comma change. I think there was a point where it was decided to have the comma after the last element in a hash.

I don't think the default configuration has ever done that. Users can configure `Style/TrailingCommaInArguments` and `Style/TrailingCommaInLiteral` to require trailing commas if they want. (`TrailingComma` has been split into 2 different cops.)

What source checksum is printed when you use the 0.35.1 gem, rather than `master`? (You will need to add that debug print statement in your gem directory.)

> Where can I see a list of the existing cops after a new version comes out?

Does anyone else have a solution for this?

**Comment by vinagrito** [comment]
>What source checksum is printed when you use the 0.35.1 gem, rather than master? (You will need to add that debug print statement in your gem directory.)

***** RuboCop source checksum: "cc718ce059f631654ce2cfedfd35bc43"

**Comment by alexdowad** [comment]
Hmm. It's a bit bewildering why you are having what appears like a problem with stale cached data. Maybe I am wrong about that? @jonas054, any ideas?

**Comment by jonas054** [comment]
@alexdowad It looks like the cached data has got the wrong format.

```json
[{"severity":{"name":"convention"},
```
instead of
```json
[{"severity":"convention",
```

We save a Severity object (https://github.com/bbatsov/rubocop/blob/master/lib/rubocop/cached_data.rb#L34) when we convert to JSON. Maybe it's not well defined how an object should be represented in JSON? Better to call `.name` on the object?

**Comment by jonas054** [comment]
The error looks very similar to #2694, which has been fixed. @vinagrito Can you check if your error can be reproduced on the current `master`?

**Comment by alexdowad** [comment]
@vinagrito Ping.

**Comment by vinagrito** [comment]
@alexdowad after pulling the current `master` the version it runs is `0.37.0`. It now behaves consistently on both cases I exposed above
>
```ruby
1  def some_method
2       super.merge(
3     a_key: {
4       another_key: "value"
5     },
6   )
7 end
```

This now also fails :)

**Comment by jonas054** [comment]
> Avoid comma after the last parameter of a method call

This is reported by `Style/TrailingCommaInArguments`, not `Style/TrailingCommaInLiteral`. By default it enfoces `no_comma`, so the behavior you describe here @vinagrito is expected. Closing.
